crypto/aes/asm/aes-s390x.pl

```diff
@@ -1405,7 +1405,42 @@ ()
 	clr	%r0,%r1
 	jl	.Lctr32_software
 
-	stm${g}	%r6,$s3,6*$SIZE_T($sp)
+	stm${g}	$s2,$s3,10*$SIZE_T($sp)
+
+	clr	$len,%r1
+	jle	.Lctr32_nokma	# kma is slower for <= 16 blocks
+
+	larl	%r1,OPENSSL_s390xcap_P
+	lr	$s2,%r0
+	llihh	$s3,0x8000
+	srlg	$s3,$s3,0($s2)
+	ng	$s3,S390X_KMA(%r1)	# check kma capability vector
+	jz	.Lctr32_nokma
+
+	aghi	$sp,-112	# prepare parameter block
+	lhi	%r1,0x0600
+	sllg	$len,$len,4
+	or	%r0,%r1		# set HS and LAAD flags
+	lmg	$s2,$s3,0($ivp)
+	ahi	$s3,-1		# kma requires counter-1
+	la	%r1,0($sp)
+	mvc	80(32,$sp),0($key)	# copy key
+	stmg	$s2,$s3,64($sp)	# copy iv
+	st	$s3,12($sp)	# copy counter
+	lghi	$s3,0		# no AAD
+
+	.long	0xb929a042	# kma $out,$s2,$inp
+	brc	1,.-4		# pay attention to "partial completion"
+
+	xc	80(32,$sp),80($sp)	# wipe key copy
```

Comment:
Well, it's not like it has to be zero-ed, it's sufficient to write some non-sensitive values, e.g. `stmg %r0,%r3,80($sp)` would do...

Reply:
right.

Reply:
xc actually would need less cycles than stmg. see other comment